mirage: init at 0.5.1

colemickens opened the pull request:
###### Motivation for this change

Package new application `mirage` as attribute `mirage-im` (there's an existing alias, maybe I could remove it?).

This is a draft since the application doesn't work yet.

This fixes #84147.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

doronbehar commented on pkgs/applications/networking/instant-messengers/mirage/default.nix:
```diff
@@ -0,0 +1,30 @@
+{ mkDerivation, stdenv, lib, fetchFromGitHub, cmake, qtbase, qtquickcontrols2
+, qtkeychain, qtmultimedia, qttools, libqmatrixclient_0_5, libQuotient, gnome3
```
libQuotient and libqmatrixclient are not used.

colemickens replied:
Yeah, sorry, left in draft so people wouldn't review too closely. I just copied quaternion so there's lots of left over cruft. I'lll clean it up once it actually does something (anything) on launch.

doronbehar reviewed (commented):
> This is a draft since the application doesn't work yet.

What happens when you run it from the command line?

cole-h commented:
@GrahamcOfBorg eval

(bot's been having a bad time recently)

ajs124 commented:
If you just want to try out this app and don't feel like finishing the package, they provide an AppImage, which you can run/package through `appimageTools.wrapType2`.

colemickens commented:
Well, I'm making some progress at least.
![output](https://user-images.githubusercontent.com/327028/82382175-61c13f00-99e0-11ea-92f9-2f5556d75cea.png)

colemickens committed:
mirage: init at 0.5.1

colemickens commented:
This is working for me now, but it needs cleanup before it's ready for review.

symphorien commented:
The update to matrix-nio breaks pantalaimon: https://github.com/matrix-org/pantalaimon/blob/master/setup.py#L32

symphorien commented:
When running I get a black window with the message:
```
! 14:55:58 | qrc:/src/gui/UI.qml:7:1: module "QtGraphicalEffects" is not installed
```

symphorien commented:
You can include this commit for compatibility of pantalaimon with matrix-nio 0.13.

https://github.com/symphorien/nixpkgs/commit/3e84e683977d542f644a00d5b1d804f6d1060269

colemickens commented:
@symphorien I wonder if that happened after a rebase as I saw this too when installed in my system profile. I added the extra qt dependency and it seems to have fixed it. I also added your pantalaimon commit. Thank you for the testing, review, and feedback.

symphorien commented:
I tried mirage and it works, except that I cannot decrypt messages sent or received in mirage. I wonder if there is an optional dependency to enable it, or something like that, as it is advertised that it should work.

For some reason, the update of matrix-nio makes pantalaimon tests hang. (the update of pantalaimon without the upgrade of matrix-nio works fine) Can you reproduce? Unfortunately I'm a bit short on time to investigate.

colemickens commented:
It segfaults, I'm concerned I'm not getting qt wrapped properly.

colemickens commented:
It looks like my derivation is fine, but Mirage doesn't handle upgrades very well currently: https://github.com/mirukana/mirage/issues/90

jonringer commented on pkgs/development/python-modules/cssselect2/default.nix:
```diff
@@ -1,28 +1,28 @@
+  src = fetchFromGitHub {
+    owner = "Kozea";
+    repo = "cssselect2";
+    rev = "${version}";
```
```suggestion
    rev = version;
```

jonringer reviewed (commented):
did you want to move all the package fixups into one PR, and mirage addition into another PR?

I could merge the fixes relatively quickly, but adding a new package is a little more involved

ZerataX commented:
I'm very new to nix packaging so I'm sorry if these questions are stupid, but reading through the [mirage install manual](https://github.com/mirukana/mirage/blob/master/INSTALL.md#manual-installation) recommends:
- libjpeg-turbo
- zlib
- libtiff
- libwebp
- openjpeg2
- libmediainfo 

and the requirements.txt also includes pyfastcopy, for which no nix package exists yet afaik.

jonringer commented:
> I'm very new to nix packaging so I'm sorry if these questions are stupid, but reading through the [mirage install manual](https://github.com/mirukana/mirage/blob/master/INSTALL.md#manual-installation) recommends:
> 
>     * libjpeg-turbo
> 
>     * zlib
> 
>     * libtiff
> 
>     * libwebp
> 
>     * openjpeg2
> 
>     * libmediainfo
> 
> 
> and the requirements.txt also includes pyfastcopy, for which no nix package exists yet afaik.

not everything is a hard requirement, and in this particular case, these dependencies are meant for the pillow package to work correctly, and that is handled by the pillow python package https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/pillow/default.nix

ZerataX commented:
ah right thanks @jonringer that's kinda what I thought. and libmediainfo is installed by pymediainfo.

And pyfastcopy is apparently only interesting for python versions < 3.8
https://github.com/mirukana/mirage/blob/87d34082deb1db3c8f784876cd4aa7202471b8db/src/backend/media_cache.py#L25

colemickens committed:
mirage: 0.5.1 -> 0.5.2

jonringer commented:
you should be able to rebase on latest master to get rid of matrix and pant changes

jonringer commented on pkgs/applications/networking/instant-messengers/mirage/default.nix:
```diff
@@ -0,0 +1,54 @@
+
+  qmakeFlags = [ "PREFIX=${placeholder "out"}" ];
+
+  postInstall = ''
+    buildPythonPath "$out $pythonPath"
+    wrapProgram $out/bin/mirage \
+      --prefix PYTHONPATH : "$PYTHONPATH" \
+      --prefix QML2_IMPORT_PATH : "${pyotherside}/${qtbase.qtQmlPrefix}"
+  '';
```
to avoid double wrapping
```suggestion
  dontWrapQtApps = true;
  postInstall = ''
    buildPythonPath "$out $pythonPath"
    wrapProgram $out/bin/mirage \
      --prefix PYTHONPATH : "$PYTHONPATH" \
      --prefix QML2_IMPORT_PATH : "${pyotherside}/${qtbase.qtQmlPrefix}" \
      ''${qtWrapperArgs[@]}
  '';
```

colemickens replied:
If I do this, I get `! 13:15:58 | QQmlComponent: Component is not ready` as the only console output.

I tried, along with `dontWrapQtApps`, adding either `''${qtWrapperArgs[@]}"` or `"''${qtWrapperArgs[@]}"` and neither worked.

jonringer replied:
did you put the `\` and the end of the previous line? I was able to get this to build and launch fine.

jonringer replied:
also, it should just be `''${qtWrapperArgs[@]}` as I've written. Want it to be expanded as part of the arguments passed to `wrapProgram`

The `''$` is to prevent nix from interpolating it, and so it should just expand as a bash array.

colemickens replied:
:( No `set -u` for the hooks I guess. I think I had spelled `wrapper` wrong...

colemickens replied:
(`"''${qtWrapperArgs[@]}"` appears in nixpkgs tree many times and expands correctly, despite how it looks. also, it works :grin: )

jonringer replied:
I'm just not quite sure why it does work.... the double quotes should be making it into a single "string". w/e, if it works, it works

jonringer commented on pkgs/applications/networking/instant-messengers/mirage/default.nix:
```diff
@@ -0,0 +1,56 @@
+    wrapProgram $out/bin/mirage \
+      --prefix PYTHONPATH : "$PYTHONPATH" \
+      --prefix QML2_IMPORT_PATH : "${pyotherside}/${qtbase.qtQmlPrefix}" \
+      "''${qtWrapperArgs[@]}"
```
I don't feel strongly about this one though, as we mentioned it above

colemickens commented on pkgs/applications/networking/instant-messengers/mirage/default.nix:
```diff
@@ -0,0 +1,56 @@
+    olm pyotherside
+  ];
+
+  propagatedBuildInputs = pypkgs;
```
If I remove this line, I get these errors and it doesn't load:

```
~ 19:21:36 | Got library name:  "/nix/store/3z4vm3gsf9zvsj9im9xmx8pmdia8cjqq-pyotherside-1.5.9/lib/qt-5.12.7/qml/io/thp/pyotherside/libpyothersideplugin.so"
~ 19:21:36 | Using Wayland-EGL
~ 19:21:36 | Using the 'xdg-shell' shell integration
~ 19:21:36 | "PyOtherSide error: Traceback (most recent call last):\n\n  File \"qrc:/src/backend/qml_bridge.py\", line 26, in <module>\n    from .pyotherside_events import CoroutineDone, LoopException\n\n  File \"qrc:/src/backend/pyotherside_events.py\", line 8, in <module>\n    from .utils import serialize_value_for_qml\n\n  File \"qrc:/src/backend/utils.py\", line 23, in <module>\n    import aiofiles\n\nModuleNotFoundError: No module named 'aiofiles'\n"
! 19:21:36 | Unhandled PyOtherSide error: Cannot import module: backend.qml_bridge (Traceback (most recent call last):

  File "qrc:/src/backend/qml_bridge.py", line 26, in <module>
    from .pyotherside_events import CoroutineDone, LoopException

  File "qrc:/src/backend/pyotherside_events.py", line 8, in <module>
    from .utils import serialize_value_for_qml

  File "qrc:/src/backend/utils.py", line 23, in <module>
    import aiofiles

ModuleNotFoundError: No module named 'aiofiles'
)
~ 19:21:36 | "PyOtherSide error: Traceback (most recent call last):\n\n  File \"<string>\", line 1, in <module>\n\nNameError: name 'BRIDGE' is not defined\n"
! 19:21:36 | Unhandled PyOtherSide error: Function not found: 'BRIDGE.call_backend_coro' (Traceback (most recent call last):

  File "<string>", line 1, in <module>

NameError: name 'BRIDGE' is not defined
)
```

colemickens replied:
going to push these commits up just so you can verify.

jonringer replied:
Yea, I was looking into a way around this, the `buildPythonPath` searches what being propagated to collect the entire python dependency tree. However, this also exports the python interpreter to the user's environment :(

colemickens commented on pkgs/applications/networking/instant-messengers/mirage/default.nix:
```diff
@@ -0,0 +1,56 @@
+, python3Packages, pyotherside
+}:
+
+let pypkgs= with python3Packages; [
+  aiofiles filetype matrix-nio appdirs cairosvg
+  pymediainfo setuptools html-sanitizer mistune blist
+  pyotherside
+]; in
```
done

colemickens commented on pkgs/applications/networking/instant-messengers/mirage/default.nix:
```diff
@@ -0,0 +1,56 @@
+{ stdenv, lib, mkDerivation, fetchFromGitHub
```
done

jonringer commented:
do you mind squashing/removing some of the fixup commits, particularly:
```
mirage: feedback
mirage: also remove propagatedBuildInputs
Revert "mirage: also remove propagatedBuildInputs" 
```

jonringer reviewed (approved):
I can't think of a way to improve this. I would like to prevent the python interpreter from getting propagated.
cc @FRidh 

otherwise, the app seems to work, and the gui launches fine
```
https://github.com/NixOS/nixpkgs/pull/84150
1 package built:
mirage-im
```

symphorien commented:
One way to prevent propagation of python is:
```diff
diff --git a/pkgs/applications/networking/instant-messengers/mirage/default.nix b/pkgs/applications/networking/instant-messengers/mirage/default.nix
index 1101d6f9b06..8e99ca64123 100644
--- a/pkgs/applications/networking/instant-messengers/mirage/default.nix
+++ b/pkgs/applications/networking/instant-messengers/mirage/default.nix
@@ -44,7 +44,10 @@ mkDerivation rec {
     wrapProgram $out/bin/mirage \
       --prefix PYTHONPATH : "$PYTHONPATH" \
       "''${qtWrapperArgs[@]}"
-    '';
+  '';
+  postFixup = ''
+    rm $out/nix-support -r
+  '';
 
   meta = with lib; {
     description = "A fancy, customizable, keyboard-operable Qt/QML+Python Matrix chat client for encrypted and decentralized communication.";
```
